### PR TITLE
Recursive Division Sphere

### PR DESCRIPTION
**Pull request by dpilch**
Created a sphere using recursive division. Sphere(origin, radius, detail)
origin - sphere origin is at the center of the sphere.
radius - used to scale the sphere
detail - the number of division used, more divisions creates a smoother appearance.

Recursive division will create a lot of faces so it is much more expensive than the other shapes.

I also added a few functions to the Vector class that I used for creating the sphere.

screen shot 2015-05-06 at 7 03 54 pm

### CODE REVIEW

**Comment by jdan** [comment]
Heya Dane :)

Mind explaining how it works? The code is a bit hard to follow.

Cool stuff though.

**Comment by NetOperatorWibby** [comment]
NICE

**Comment by dpilch** [comment]
It is very complicated, I should have done more commenting.
The basic idea is that it starts with pyramid with a triangle as a base. It then divides each of these triangular faces recursively. 
```javascript
divideTriangle(v1, v2, v3, numDivisions, sphere);
divideTriangle(v4, v3, v2, numDivisions, sphere);
divideTriangle(v1, v4, v2, numDivisions, sphere);
divideTriangle(v1, v3, v4, numDivisions, sphere);
```

Each triangle is divided by getting three new points. These new points are roughly the midpoints between the two points plus a bit further out radially to shape the sphere.
```javascript
var v1d = (Vector.add(a,b)).normalize();
var v2d = (Vector.add(a,c)).normalize();
var v3d = (Vector.add(b,c)).normalize();
```
img_0523
Now it looks kinda like the triforce but the middle triangle is raised a little. Connecting all these points then leaves 4 triangles that are then divide the same way until the number of division reaches zero and triangle is added as face.
```javascript
var face = new Path([Vector.toPoint(Vector.add(Vector.mult(a,radius), offset)),
                                 Vector.toPoint(Vector.add(Vector.mult(b,radius), offset)),
                                 Vector.toPoint(Vector.add(Vector.mult(c,radius), offset))]);
sphere.push(face);
```
The offset and radius are used to translate and scale the sphere. This method would always create a sphere at the origin with a radius of 1 otherwise.
screen shot 2015-05-07 at 4 27 23 pm

**Comment by jdan** [comment]
Sorry it took so long for me to get back to you.

While I appreciate the effort and the explanation of your code, I've put a freeze on all shapes for now. I'm more focused on rewriting Isomer to use a WebGL backend so I'm not accepting any PRs for new shapes at this time.

I'd encourage you to turn this into your own module and simply use it in your code alongside isomer. This would make for a cool blog post - is this your own algorithm for drawing spheres?

```javascript
var Isomer = require("isomer");
var Sphere = require("isomer-recursive-divison-sphere");
var iso = new Isomer(...); 

iso.add(new Sphere(...));
```
